packages/cactus-plugin-satp-hermes/src/main/typescript/core/recovery/rollback/stage3-rollback-strategy.ts

```diff
@@ -0,0 +1,136 @@
+import { Logger, LoggerProvider } from "@hyperledger/cactus-common";
+import { SATPSession } from "../../satp-session";
+import { RollbackStrategy } from "./rollback-strategy-factory";
+import {
+  RollbackLogEntry,
+  RollbackState,
+} from "../../../generated/proto/cacti/satp/v02/crash_recovery_pb";
+import { SATPBridgesManager } from "../../../gol/satp-bridges-manager";
+import { ILocalLogRepository } from "../../../repository/interfaces/repository";
+
+export class Stage3RollbackStrategy implements RollbackStrategy {
+  private log: Logger;
+  private bridgesManager: SATPBridgesManager;
+  private logRepository: ILocalLogRepository;
+
+  constructor(
+    bridgeManager: SATPBridgesManager,
+    localLog: ILocalLogRepository,
+  ) {
+    this.log = LoggerProvider.getOrCreate({ label: "Stage3RollbackStrategy" });
+    this.bridgesManager = bridgeManager;
+    this.logRepository = localLog;
+  }
+
+  async execute(session: SATPSession): Promise<RollbackState> {
+    const fnTag = "Stage3RollbackStrategy#execute";
+    this.log.info(`${fnTag} Executing rollback for Stage 3`);
+
+    const sessionData = session.hasClientSessionData()
+      ? session.getClientSessionData()
+      : session.getServerSessionData();
+
+    if (!sessionData) {
+      throw new Error(`${fnTag}, session data is not correctly initialized`);
+    }
+
+    const network = sessionData.senderGatewayNetworkId;
+
+    if (!network) {
+      throw new Error(
+        `${fnTag}: Unable to determine network from session data.`,
+      );
+    }
+
+    const bridgeManager = this.bridgesManager.getBridge(network);
+
+    const rollbackState = new RollbackState({
+      sessionId: session.getSessionId(),
+      currentStage: String(sessionData.hashes?.stage3),
+      stepsRemaining: 2,
+      rollbackLogEntries: [],
+      estimatedTimeToCompletion: "",
+      status: "IN_PROGRESS",
+      details: "",
+    });
+
+    try {
+      const assetId = sessionData.senderAsset?.tokenId;
+      const amount = sessionData.senderAsset?.amount;
+      if (!assetId) {
+        throw new Error(`${fnTag}: Asset ID is undefined`);
+      }
+
+      if (!amount) {
+        throw new Error(`${fnTag}, Amount is missing`);
+      }
+
+      this.log.info(`${fnTag} Asset Id: ${assetId} amount: ${amount}`);
+
+      await bridgeManager.burnAsset(assetId, Number(amount));
```

Comment:
In the Stage 3 rollback, the rollback is only feasible if it occurs before the asset is minted on the receiver chain. Once minting happens, if the gateway encounters an issue and fails to assign the minted amount to the recipient, a rollback can no longer be initiated. Is this the reason why a rollback isn’t considered in such cases?

Wouldn't it make sense, then, for the minted amount on the receiver chain to be burned and re-minted on the source chain too?

Reply:
@RafaelAPB